fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -551,82 +515,68 @@ def _setup_bucket_strategy(self) -> None:
         # - these are only the trainable parameters, so they should produce grads
         # - they are sorted by increasing size
         self.buckets = {}
+        self._should_bucket_grad = []
+        self._should_bucket_grad = [False for _ in range(len(self._trainable_params))]
 
-        for param in self._trainable_params:
+        for index, param in enumerate(self._trainable_params):
             device = param.device
             dst_rank = self._trainable_param_to_rank[param]
 
             if param.device not in self.buckets.keys():
-                self.buckets[param.device] = [
```

Comment:
this could lead to a subtle over provision, if not all ranks where present for all devices. Moving that to allocating the bucket on demand, when it's really required